app/resources/v1/geoip.js

```diff
@@ -0,0 +1,101 @@
+const config = require('config')
+const request = require('request')
+const logger = require('../../../lib/logger.js')()
+const redis = require('redis')
+const url = require('url')
+
+const IP_GEOLOCATION_TIMEOUT = 500
+
+exports.get = function (req, res) {
+  if (req.headers.host !== config.app_host_port) {
+    res.status(403).json({ status: 403, error: 'Not allowed to access API' })
+    return
+  }
+
+  const requestGeolocation = function (isRedisConnected = true) {
+    console.log('requesting geolocation')
+    const url = `${config.geoip.protocol}${config.geoip.host}?access_key=${config.geoip.api_key}`
+
+    request(url, { timeout: IP_GEOLOCATION_TIMEOUT }, function (error, response, body) {
+      if (error) {
+        logger.error(error)
+        // Error codes 101 and 104 are provided by ipstack to specify invalid/missing api key
+        // and maximum monthly API requests reached, respectively. Since ipstack already has
+        // error messages included with these errors, just send the message to the front-end.
+        if (error.code === 101 || error.code === 104) {
+          res.status(error.code).json({ status: error.code, error: error.info })
+        } else if (error.code === 'ETIMEDOUT') {
+          // If request takes longer than declared IP_GEOLOCATION_TIMEOUT, return status code 408
+          res.status(408).json({ status: 408, error: 'Request timed out' })
+        }
+        return
+      }
+
+      if (isRedisConnected && req.ip) {
+        client.set(req.ip, body, redis.print)
+      }
+
+      res.status(200).send(body)
+    })
+  }
+
+  const handleRedisErrors = function (error, isRedisConnected = true) {
+    // If any errors involving Redis, log the error and then request geolocation
+    // directly from ipstack.
+    logger.error(error)
+    if (!isRedisConnected) {
+      client.end(true)
+    }
+    requestGeolocation(isRedisConnected)
+  }
+
+  const authenticateRedis = function () {
```

Comment:
We should document that we're wrapping `redis` functionality to make it a Promise.

Reply:
Can we use Node's built in `util.promisify()` to do this for us? https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original